src/topology/metric_space/basic.lean

```diff
@@ -1349,3 +1350,78 @@ le_trans (diam_mono ball_subset_closed_ball bounded_closed_ball) (diam_closed_ba
 end diam
 
 end metric
+
+/- Metric spaces are often defined as quotients of spaces endowed with a "distance"
+  function satisfying the triangular inequality, but for which `dist x y = 0` does
+  not imply x = y. We call such a space a premetric space.
+  `dist x y = 0` defines an equivalence relation, and the quotient
+  is canonically a metric space. -/
+
+class premetric_space (α : Type u) extends has_dist α : Type u :=
+(dist_self : ∀ x : α, dist x x = 0)
+(dist_comm : ∀ x y : α, dist x y = dist y x)
+(dist_triangle : ∀ x y z : α, dist x z ≤ dist x y + dist y z)
+
+namespace premetric
+protected lemma dist_nonneg {α : Type u} [premetric_space α] {x y : α} : 0 ≤ dist x y :=
+begin
+  have := calc
+    0 = dist x x : (premetric_space.dist_self _).symm
+    ... ≤ dist x y + dist y x : premetric_space.dist_triangle _ _ _
+    ... = dist x y + dist x y : by simp [premetric_space.dist_comm],
+  by linarith
+end
+
+/-- The canonical equivalence relation on a premetric space. -/
+instance dist_setoid (α : Type u) [premetric_space α] : setoid α :=
```

Comment:
This defines a setoid on each `α` which happens to have a `premetric_space`. I think this should be a `def` with `local attribute [instance] dist_setoid`

Reply:
I'm more and more worried about too much generic type class instances. This instance is always triggered when a `setoid` is looked up, even if no `premetric_space` is found. We have many instances which look like this, but I think we might be more careful with adding them. Maybe Lean 4 has better caching again. But currently I would like to avoid it.

So yes, while it is a valid `instance`, please change it to a `def` with `local attribute ...`

Reply:
Done